File: bindings/python/tests/test_conformance.py
from __future__ import annotations

import json
from pathlib import Path

import kairo_ecs

FIXTURES = Path(__file__).resolve().parents[3] / "conformance" / "fixtures"


def _load_fixture(name: str) -> dict[str, object]:
    with (FIXTURES / name).open(encoding="utf-8") as handle:
        return json.load(handle)


def _dispatch_all(scheduler: kairo_ecs.Scheduler) -> list[int]:
    kinds: list[int] = []
    while True:
        outcome, event = scheduler.step()
        if outcome is kairo_ecs.StepOutcome.EMPTY:
            return kinds
        assert outcome is kairo_ecs.StepOutcome.DISPATCHED
        assert event is not None
        kinds.append(event.kind)


def test_deterministic_ordering():
    fixture = _load_fixture("deterministic_ordering.json")
    scheduler = kairo_ecs.Scheduler()

    for event in fixture["events"]:
        assert isinstance(event, dict)
        scheduler.schedule_at(
            int(event["at_ticks"]),
            priority=int(event["priority"]),
            kind=int(event["kind"]),
        )

    assert fixture["version"] == 1
    assert _dispatch_all(scheduler) == fixture["expected_kind_order"]


def test_cancellation():
    fixture = _load_fixture("cancellation.json")
    scheduler = kairo_ecs.Scheduler()

    for event in fixture["events"]:
        assert isinstance(event, dict)
        event_id = scheduler.schedule_at(
            int(event["at_ticks"]),
            priority=int(event["priority"]),
            kind=int(event["kind"]),
        )
        if event.get("cancel") is True:
            assert scheduler.cancel(event_id) is True

    assert _dispatch_all(scheduler) == fixture["expected_kind_order"]


def test_zero_delay_guard():
    fixture = _load_fixture("zero_delay_guard.json")
    scheduler = kairo_ecs.Scheduler()

    for event in fixture["events"]:
        assert isinstance(event, dict)
        scheduler.schedule_at(
            int(event["at_ticks"]),
            priority=int(event["priority"]),
            kind=int(event["kind"]),
        )

    assert _dispatch_all(scheduler) == fixture["expected_kind_order"]


def test_rng_replay():
    fixture = _load_fixture("rng_replay.json")

    assert fixture["run_seed"] == 7
    assert len(fixture["expected_stream"]) == 4

File: bindings/r/tests/testthat/test-conformance.R
fixture_path <- function(name) {
  roots <- normalizePath(getwd(), mustWork = FALSE)
  for (i in seq_len(6)) {
    roots <- c(roots, normalizePath(file.path(tail(roots, 1), ".."), mustWork = FALSE))
  }

  candidates <- file.path(
    unique(roots),
    "conformance",
    "fixtures",
    name
  )
  candidates[file.exists(candidates)][1]
}

read_fixture <- function(name) {
  testthat::skip_if_not_installed("jsonlite")
  path <- fixture_path(name)
  if (is.na(path)) {
    stop(paste("missing fixture", name), call. = FALSE)
  }
  jsonlite::read_json(path)
}

run_scheduler_fixture <- function(fixture_name) {
  fixture <- read_fixture(fixture_name)
  scheduler <- kairoecs_new_scheduler(run_id = fixture$fixture, time_scale = "ticks")

  scheduled <- list()
  for (event in fixture$events) {
    scheduler <- kairoecs_schedule_at(
      scheduler,
      time_ticks = event$at_ticks,
      event_kind = as.character(event$kind),
      priority = event$priority,
      payload_ref = paste0("fixture:", fixture$fixture)
    )
    scheduled[[as.character(event$kind)]] <- scheduler$next_event_id - 1
  }

  for (event in fixture$events) {
    if (isTRUE(event$cancel)) {
      scheduler <- kairoecs_cancel_event(scheduler, scheduled[[as.character(event$kind)]])
    }
  }

  scheduler <- kairoecs_run_until(scheduler, Inf)
  log <- kairoecs_event_log(scheduler)
  dispatched <- log[log$status == "dispatched", , drop = FALSE]

  as.integer(dispatched$event_kind)
}

test_that("deterministic ordering fixture drives the scheduler facade", {
  expect_identical(
    run_scheduler_fixture("deterministic_ordering.json"),
    c(1L, 2L, 4L, 3L)
  )
})

test_that("cancellation fixture drives the scheduler facade", {
  expect_identical(
    run_scheduler_fixture("cancellation.json"),
    c(1L, 3L)
  )
})

test_that("zero-delay guard fixture keeps deterministic priority order", {
  expect_identical(
    run_scheduler_fixture("zero_delay_guard.json"),
    c(1L, 2L, 5L, 10L)
  )
})

test_that("rng_replay fixture remains metadata-only until the R RNG facade exists", {
  fixture <- read_fixture("rng_replay.json")
  expect_equal(fixture$run_seed, 7)
  expect_length(fixture$expected_stream, 4)
  expect_equal(fixture$requirement, "same run seed and entity handle produce the same stream")
})
